Blog: Maybe in λ-Calculus (#3)
bash/bash.github.io: content/blog/2023-04-06-lc-maybe.md

+++
title = "Maybe in λ-calculus"
description = "Read along as I implement the Maybe type and some useful functions for working on Maybe values in λ-calculus."
draft = true
authors = ["Ruben"]
+++

I'm currently taking a class on [Models of Computation]. If you're studying at the University of Zurich or ETHZ I highly recommend that you take this class :) You'll get to solve really cool puzzles each week as part of the class.

> My favourite model of computation so far is [Lambda Calculus]. A [functional programming enthusiast], I may be a bit biased though…

Reviewer: "As a" instead of "A"

Author: eee5a9df53dde569f4761c42efb75652c155ad49

Going forward, I will assume that you have some basic knowledge of [Lambda Calculus].

## What is a Maybe?
A `Maybe` value represents an optional value. It can either be present (which we will call `Just`) or absent (which we will call `Nothing`).

> In some programming languages, such as [Rust], `Maybe` is called `Option` instead.

Reviewer: you could add that Java has the Optional type

Author: d0252b37a171199d73b3db7b1f485013e0e4bb4f

Don't worry, I will not be spoiling any of the exercises from the [Models of Computation] class ;)

### Church Numerals
Church Numerals encode a given natural number `n`, by applying a function `n` times.

The first function application needs some starting value (or *seed*),
so each number is a function that takes two arguments:

```λ
λf.λx.EXPR
```

Here are the expressions for the numbers zero to three:
(I'm using `#` to denote comments)
```λ
λf.λx.x              # Zero
λf.λx.f x            # One
λf.λx.f (f x)        # Two
λf.λx.f (f (f x))    # Three
```

### Tuples
Tuples are similar to Church Numerals in that tuple values accept some function that they then apply.
In the case of tuples, the function is applied with two arguments: one for the first value of the tuple and one for the second value of the tuple.

So the tuple `(X, Y)` would correspond with the following λ expression:
```λ
λs.s X Y
```

We can then generalize this to an expression that creates a tuple:
```λ
λa b.λs.s a b
# Note that this is equivalent to a function with three arguments:
λa b s.s a b
# ... or three functions with one argument each:
λa.λb.λs.s a b
```

> Picking out the first or second value is easy: We call our tuple value with a function that takes two arguments and returns one and discards the other.

Author: f6527ba80b68c88e6d049b4a5b5ffe5136405fe8

So picking the first value out of a tuple would look something like this:

> ```λ

Reviewer: you legend <3

Author: Thaaanks :)

#   ─────────┴──────────
=> (λm f.bind m (lift f)) (Just X) TIMES_TWO
#        ─┬──    ─┬──
#         │       ╰────────────────────╮
#         ╰──────╮                     │
#         ───────┴────────      ───────┴────────
=> (λm f.(λm f.m f Nothing) m ((λf.λx.Just (f x)) f)) (Just X) TIMES_TWO
=> (λm f.m ((λf.λx.Just (f x)) f) Nothing) (Just X) TIMES_TWO
=> (λm f.m (λx.Just (f x)) Nothing) (Just X) TIMES_TWO
=> (Just X) (λx.Just (TIMES_TWO x)) Nothing
#   ─┬──        ─┬──                ───┬───
#    │           ╰──────────────╮      ╰────────────────────╮
#    ╰────╮                     │                           │
#    ─────┴──────          ─────┴──────                  ───┴───
=> ((λa.λj n. j a) X) (λx.(λa.λj n. j a) (TIMES_TWO x)) (λj n. n)
=> (λj n. j X) (λx.λj n. j (TIMES_TWO x)) (λj n. n)
=> (λx.λj n. j (TIMES_TWO x)) X
=> λj n. j (TIMES_TWO X)
```

This is the same as `Just (TIMES_TWO X)`. Yayy it worked ✨
We could do a similar demonstration for `Nothing`, but I leave that as an exercise to you dear reader.

<!-- TODO: filter (with example, use IS_EVEN as predicate) -->

## Outlook
One of my favourite properties of λ-calculus is the composable nature of expressions. \

> We could easily build more complicated structures (trees, linked lists, etc.) from the functions discussed in this post. I might write a follow-up post soon :)

Reviewer: I might write a follow-up post soon :) <- maybe don't say soon, that's a classic :) what about `I might write a follow-up to investigate that`

Reviewer: or smth like that, just not soon :P

Author: 60c3a3104a98ce70a035ff3b901c382ecd081efe

Don't worry, I will not be spoiling any of the exercises from the [Models of Computation] class ;)

### Church Numerals
Church Numerals encode a given natural number `n`, by applying a function `n` times.

The first function application needs some starting value (or *seed*),
so each number is a function that takes two arguments:

```λ
λf.λx.EXPR
```

Here are the expressions for the numbers zero to three:
(I'm using `#` to denote comments)
```λ
λf.λx.x              # Zero
λf.λx.f x            # One
λf.λx.f (f x)        # Two
λf.λx.f (f (f x))    # Three
```

### Tuples
Tuples are similar to Church Numerals in that tuple values accept some function that they then apply.
In the case of tuples, the function is applied with two arguments: one for the first value of the tuple and one for the second value of the tuple.

So the tuple `(X, Y)` would correspond with the following λ expression:
```λ
λs.s X Y
```

We can then generalize this to an expression that creates a tuple:
```λ
λa b.λs.s a b
# Note that this is equivalent to a function with three arguments:
λa b s.s a b
# ... or three functions with one argument each:
λa.λb.λs.s a b
```

> Picking out the first or second value is easy: We call our tuple value with a function that takes two arguments, returns one and discards the other.

Author: ```suggestion
Picking out the first or second value is easy: We call our tuple value with a function that picks one of its two arguments.
```
